src/catalog/src/durable/impls/persist.rs

```diff
@@ -828,31 +856,56 @@ impl DurableCatalogState for PersistCatalogState {
-        // If the transaction is empty then we don't error, even in read-only mode. This matches the
-        // semantics that the stash uses.
-        if !txn_batch.is_empty() && self.is_read_only() {
-            return Err(DurableCatalogError::NotWritable(
-                "cannot commit a transaction in a read-only catalog".to_string(),
-            )
-            .into());
-        }
+        async fn commit_transaction_inner(
+            catalog: &mut PersistCatalogState,
+            txn_batch: TransactionBatch,
+        ) -> Result<(), CatalogError> {
+            // If the transaction is empty then we don't error, even in read-only mode. This matches the
+            // semantics that the stash uses.
+            if !txn_batch.is_empty() && catalog.is_read_only() {
+                return Err(DurableCatalogError::NotWritable(
+                    "cannot commit a transaction in a read-only catalog".to_string(),
+                )
+                .into());
+            }
 
-        let current_upper = self.upper.clone();
-        let next_upper = current_upper.step_forward();
+            let current_upper = catalog.upper.clone();
+            let next_upper = current_upper.step_forward();
 
-        let updates = StateUpdate::from_txn_batch(txn_batch, current_upper);
-        debug!("committing updates: {updates:?}");
+            let updates = StateUpdate::from_txn_batch(txn_batch, current_upper);
+            debug!("committing updates: {updates:?}");
 
-        if matches!(self.mode, Mode::Writable) {
-            self.compare_and_append(updates, current_upper, next_upper)
-                .await?;
-            debug!("commit successful, upper advanced from {current_upper:?} to {next_upper:?}",);
-            self.sync(next_upper).await?;
-        } else if matches!(self.mode, Mode::Savepoint) {
-            self.apply_updates(updates)?;
-        }
+            if matches!(catalog.mode, Mode::Writable) {
+                catalog
+                    .compare_and_append(updates, current_upper, next_upper)
+                    .await?;
+                debug!(
+                    "commit successful, upper advanced from {current_upper:?} to {next_upper:?}",
+                );
+                catalog.sync(next_upper).await?;
+            } else if matches!(catalog.mode, Mode::Savepoint) {
+                catalog.apply_updates(updates)?;
+            }
 
-        Ok(())
+            Ok(())
+        }
+        let counter = self
+            .metrics
+            .transaction_commit_latency_duration_seconds
+            .clone();
+        let res = commit_transaction_inner(self, txn_batch)
+            .wall_time()
+            .inc_by(counter)
+            .await;
+        self.metrics.transactions_committed.inc();
+        if let Err(e) = &res {
+            let cause = e.to_string_with_causes();
+            self.metrics
+                .transaction_commit_errors
+                .with_label_values(&[&cause])
```

Comment:
Any idea what the cardinality is here?

Reply:
Not really, is a high cardinality bad? If so, I think it's probably fine to remove the label.

Reply:
Yeah, each unique value for the label ends up being a counter in prometheus